### DISCUSSION TITLE
Inquiry about the Specific Implementation of thrust::sort

### DISCUSSION THREAD

**Discussion by GriffithLin** [NVIDIA/cccl#570] [Thrust]
I'm interested in understanding the specific implementation details of the thrust::sort function in the Thrust library. I've explored the Thrust repository on GitHub, but I couldn't locate the actual source code for the thrust::sort algorithm.

Could someone kindly provide me with information about where I can find the full source code for thrust::sort? Knowing the underlying implementation would be very helpful for my understanding of the library.

Thank you in advance for your assistance!

**Comment by pauleonix** [comment]
The implementations for the CUDA backend come from CUB. [Depending on the type of the elements to be sorted](https://github.com/NVIDIA/cccl/blob/a8971b49b35a1c604f81301c6d5549d4f01af8a5/thrust/thrust/system/cuda/detail/sort.h#L389-L485), it can use radix sort or merge sort.

**Reply by gevtushenko** [reply]
@pauleonix is right, thrust switches between merge or radix sort based on the element type. To clarify a few details:
- primitive types like char, int, float are sorted using radix sort when you use default comparison operators (`thrust::less`, `thrust::greater`)
- if you sort user-defined types or provide a custom comparison operator, thrust fallbacks to merge sort

Here are the entry points to the:
1. merge sort [implementation](https://github.com/NVIDIA/cccl/blob/main/cub/cub/device/device_merge_sort.cuh)
2. radix sort [implementation](https://github.com/NVIDIA/cccl/blob/main/cub/cub/device/device_radix_sort.cuh)
